### ISSUE TITLE
Deflate followed by inflate yields data very different to the original input

### ISSUE THREAD

**Issue by bergr01** [nodeca/pako#78]
__upd:__ https://github.com/nodeca/pako-test - bug's demo repo

---

Hi,

For one of my files during testing (it's a 93 MB heap dump file), I noticed that when the zlib.deflate is called the data size goes from 97186971 to 5772270, however if I immediately follow that with a call to zlib.inflate the result is 67615664 - so almost 30 megabytes of data seem to get lost somewhere. Is there some known issue of problems with particular files?

regards,
Greg

**Comment by puzrin** [comment]
No ideas. This library is used wide in many projects, i'm not sure that bug is here. Try to create a more simple example. Also make sure you use proper encoding if work with binary strings & if send data via XHR.

**Comment by bergr01** [comment]
Let me clarify. Here is my simple example:
  var zlib = require('pako')
  var data = <my data, doesn't matter where it comes from>

  var origLength = data.length
  console.warn('Deflating ', data.length)
  var deflatedData = zlib.deflate(data)
  console.warn('Deflated to ', deflatedData.length)
  var tmpInflate = zlib.inflate(deflatedData)
  console.warn('TmpInflation produced', tmpInflate.length, 'original length was', origLength)

yields the following in javascript console:

Deflating  97186971
Deflated to  5772270
TmpInflation produced 67615664 original length was 97186971

if the bug is not in this library, where is it?

**Comment by puzrin** [comment]
What is the type of `data` var? Try to cast into Uint8Array for sure, prior to process & measure.

If problem still exists - try to slice minimal possible peace of your data, where bug is still reproducible and post it into test repo with your testing code. We will debug and fix then.

**Comment by bergr01** [comment]
Data is already Uint8Array. Will do some slicing and let you know if I can find a small data segment that triggers the issue.

**Comment by bergr01** [comment]
Please see the attached file containing a 2.1 MB slice from my original 93 MB file, which seems to exhibit the problematic behaviour - it's not a .zip file as such, it's just a binary blob (application/octet-stream) that you can use to hopefully reproduce the issue 

[test.zip](https://github.com/nodeca/pako/files/318110/test.zip)

**Comment by puzrin** [comment]
Thanks!. May be you could create a test repo and commit there sample + js code? That will be more convenient for all IMHO.

/cc @andr83

**Comment by bergr01** [comment]
I'm sorry but I really don't have any more time to spend in this at this moment as I'm trying to get a release of our product out the door - I just wanted to report this as it had me worried for a while that the issue was elsewhere in my code, so figured I'd report it.

**Comment by puzrin** [comment]
No problem. I'll prepare it myself and update status. Thanks for your help.

**Comment by puzrin** [comment]
I confirm bug in deflate :(

https://github.com/nodeca/pako-test - created demo repo and added link to the first post.

**Comment by bergr01** [comment]
Thanks for the update Vitaly - please let me know when the fix is available, I'll be happy to re-test with my data.

**Comment by puzrin** [comment]
@bergr01 could you add more sample files (initial 90mb will be ok too)? Need to test that problem is outside of "last lost chunk". Current one produce too small difference (< 32K), and on the tail only. That's enougth to continue but can be not enougth to guarantee that future fix will resolve all possible problems.

Is 2 weeks deadline ok? Such serious problems are fixed with high priority. @andr83 said he will be back next week to take a look. If not, i'll ask one of my programmer to debug.

**Comment by bergr01** [comment]
@puzrin My original problematic file has been uploaded to dropbox, let me know if you have trouble downloading it via this link: https://www.dropbox.com/s/99500kixcckzw0c/heapdump-1451986453566.o2c.index?dl=0
This bug will not be delaying out release since it has only happened for one file so far, but we will update to a newer version of the library once the fix is available and has passed testing.

**Comment by puzrin** [comment]
Thanks. Added to test repo.

```
Deflating 97185208 bytes
=======================
Deflate by pako: 5772394 bytes
Deflate by zlib: 7674515 bytes
first difference at 5772383
----------------------
Inflate by pako(pako): 67619826 bytes
Inflate by zlib(pako): 67619826 bytes
----------------------
Inflate by pako(zlib): 97185208 bytes
Inflate by zlib(zlib): 97185208 bytes
```

Good news is that ports are ok and problem is in wrapper (it does not produce wrong stream, but terminates by wrong conditions)

**Comment by bergr01** [comment]
Thanks for the update Vitaliy, let me know when there are some other good news to report (like fixed wrapper!) :) But as I mentioned earlier, it's not a severity 1 for us at this stage, so whenever time allows.

**Comment by bergr01** [comment]
Hi Vitali, any updates on this issue? Or are you still waiting for @andr83 to return?

**Comment by puzrin** [comment]
In progress. Bug is somewhere very deep and needs more time for defug :(

**Comment by bergr01** [comment]
No worries - thanks for the update and best of luck with finding it!

**Comment by thenickdude** [comment]
I suspect this is the cause of a bunch of my users reporting their drawing files becoming damaged on my drawing tool [ChickenPaint](https://github.com/thenickdude/chickenpaint), which uses Pako to compress the RGBA pixels of the image layers. These users always have the PNG flattened version of the drawing save successfully (which doesn't use Pako), but the Pako-compressed layers are sometimes truncated (say, from 2MB -> 40kB). 

Unfortunately, I don't have any uncompressed data to test with, since the drawings are already destroyed by the time I get a complaint about them, and the problem has never popped up for me while saving my own drawings. But I'll do my best to investigate using the sample that @bergr01 has provided.

The uncompressed layers are anywhere from 5 to 100MB of very highly-compressible data (most layers are mostly fully transparent), so I might be hitting cases that haven't seen much testing before.

**Comment by puzrin** [comment]
Active fixing is in progress last 2 days. @Kirill89 reduced breaking sample to 1.5K and is debugging JS & C sources to find the difference. I hope problem will be resolved soon.

If you wish - add your samples to check.

**Comment by puzrin** [comment]
https://github.com/nodeca/pako/commit/a529410558eec6e0f22580516f6dc86cd58df591#diff-ffc028c4c0684e7a8aa7d47c4b7b8c2cR1376

Preliminary fix ready. Will recheck and publish tomorrow.

**Comment by bergr01** [comment]
Thanks for the update @puzrin

**Comment by thenickdude** [comment]
I found one of my images which fails to deflate properly with the old Pako at compression level 7 (no problem at level 6), here's the repo with the test image and modified test:

https://github.com/thenickdude/pako-issue-78

This image deflates fine with the proposed patch, thanks!

**Comment by thenickdude** [comment]
I've now tested the patched version on 10,000 of my images (4.6GB total deflated size) with no errors.

**Comment by puzrin** [comment]
Thanks for info. I'll do release today. Need to recheck code changes for sure.

**Comment by puzrin** [comment]
Released 1.0.2. Also released as 0.2.9 to give change to older dependents.

Thanks for your help!

**Comment by bergr01** [comment]
Thanks @puzrin!
